rcl/src/rcl/rmw_implementation_identifier_check.c

```diff
@@ -63,16 +89,54 @@ INITIALIZER(initialize) {
     )
     exit(ret);
   }
-  // If the environment variable is set, and it does not match, print a warning and exit.
-  if (strlen(expected) > 0 && strcmp(rmw_get_implementation_identifier(), expected) != 0) {
+  if (strlen(asserted_rmw_impl_env) > 0) {
+    // Copy the environment variable so it doesn't get over-written by the next getenv call.
+    asserted_rmw_impl = (char *)allocator.allocate(
+      strlen(asserted_rmw_impl_env) + 1, allocator.state);
+    if (!asserted_rmw_impl) {
+      RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "allocation failed")
+      exit(RCL_RET_BAD_ALLOC);
+    }
+    memcpy(asserted_rmw_impl, asserted_rmw_impl_env, strlen(asserted_rmw_impl_env) + 1);
+  }
+
+  // If both environment variables are set, and they do not match, print a warning and exit.
```

Comment:
Nit: as failing this check results in a call to exit, that's a fatal error, not a warning